Move web-client into a webworker to keep the main thread responsive

sisou opened the pull request:
## What's in this pull request?

Various changes, but all relatively minor, with minimal effect on the public API. 

> **Note**
> This PR is best reviewed commit-by-commit, as the complete change-set might be overwhelming. I made every logical change into its own commit, with expansive descriptions in the commit messages.

If anything is unclear, please ask!

#### This fixes UI freezing in browsers while the client is using lots of CPU.

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

sisou committed:
Refactor parsing in from_any methods of Address and Transaction

In Transaction, the method could not parse strings, because trying to cast the input into a PlainTransaction already returned early with an Err, preventing string parsing to run at all.

sisou committed:
Prepare client to run in a web worker

- The wasm-timer crate was incompatible, as it required access to window. I patched it and set the updated version to be used globally, as the libp2p-websocket-transport package, which is an indirect dependency, is also using it.
- Use wasm_timer::Delay for timeouts, instead of a custom implementation with js_sys promise and window.setTimeout.
- Chromium-based browsers do not emit offline/online events in workers, so I changed them to listen for postMessage messages from the main thread instead. The sender of those events will be in the main-thread index.js file in a later commit.

sisou committed:
Add missing js_name and docs for client.getTransaction

sisou committed:
Update client config defaults to testnet and matching seed URL

sisou committed:
Avoid passing WASM-memory pointers between main thread and worker

JS objects returned by wasm-bindgen'd structs are just pointers into WASM memory ({ptr: number}) with mocked class methods and properties. Since calls into in the same thread WASM are sync, users of these objects don't notice a difference.

However, proxying these structs (which **only live** in WASM memory) over to the main thread from the webworker makes all accesses async, even direct property access. That means getting e.g. `peerInfo.address` returns a `Promise<string>` instead of a `string`. Unfortunately, there is no way to represent this behavior in Typescript files, meaning users get no indication of it.

Passing plain values to JS instead copies the primitives over into JS memory, and then over to the main thread, where they are cloned and can be accessed like normal. If the struct has no methods anyway, it thus makes sense to just return a plain JS object.

For the other direction, the issue is that the object the user wants to pass from the main thread into the client in the webworker only exists in the main thread WASM memory. The client WASM in the webworker has no knowledge of it and cannot access it, especially not over the async worker barrier. This means data has to be serialized and copied over into the worker instead, which we here just simply make explicit, instead of implicit with unknown failure cases.

sisou committed:
Use a plain client config passed from main thread to worker

Similar to the last commit, we need to serialize the config to pass it from the main thread to the worker, which mean introducing a plain config version into which the config gets encoded in the main thread and decoded from in the worker.

Furthermore, this change has the advantage that `Client.create()` now simply takes a plain object with all properties optional, making it possible to pass a config object with the right properties in directly, without _having_ to go through the ClientConfiguration builder.

sisou committed:
Convert TransactionBuilder into a standalone static class

We cannot use TransactionBuilder client-connected, as the client and the TransactionBuilder (or at least the transactions that it outputs) live in different thread WASM memories, and accessing them from each other between these contexts is not (easily) possible.

Unfortunately, this means for now users have to specify the validity start height and network ID manually, for which I added accessor methods on the client.

sisou committed:
All public client methods have to be typed as async

Even if they aren't internally. This is because the communication from main thread to webworker can only be async. Typing it async in rust generates the correct types in Typescript.

sisou committed:
Separate client and client configuration structs into their own modules

In preparation for the introduction of features in the next commit.

sisou committed:
Introduce features to split client between primitives and worker

This way the primitives-only (offline) build is much smaller (unoptimized 7.0MB -> 810kB)

sisou committed:
Update package README

sisou committed:
Introduce more feature guards to reduce code duplication between build targets

Making both feature builds smaller.

sisou committed:
Resolve clippy errors and warnings

jsdanielh commented on web-client/src/address.rs:
```diff
@@ -92,9 +92,11 @@ impl Address {
     pub type AddressAnyType;
 
-    #[wasm_bindgen(typescript_type = "(Address | string)[]")]
+    // #[wasm_bindgen(typescript_type = "(Address | string)[]")]
```
Maybe remove the comment?

sisou replied:
Fixed by allowing the class instance as argument instead =)

jsdanielh commented on web-client/src/address.rs:
```diff
@@ -92,9 +92,11 @@ impl Address {
 #[wasm_bindgen]
 extern "C" {
-    #[wasm_bindgen(typescript_type = "Address | string")]
+    // #[wasm_bindgen(typescript_type = "Address | string")]
```
Maybe remove the comment?

jsdanielh commented on web-client/src/transaction.rs:
```diff
@@ -882,6 +882,7 @@ extern "C" {
     pub type PlainTransactionReceiptArrayType;
 
-    #[wasm_bindgen(typescript_type = "Transaction | PlainTransaction | string")]
+    // #[wasm_bindgen(typescript_type = "Transaction | PlainTransaction | string")]
```
Maybe remove the comment?

jsdanielh commented on web-client/src/lib.rs:
```diff
@@ -151,10 +162,46 @@ impl ClientConfiguration {
+    //         Ok(client) => client,
+    //         Err(_) => unreachable!(),
+    //     }
+    // }
```
Maybe remove the commented code?

sisou replied:
I am keeping it, but adding a TODO comment that explains that this method should be tried to be reactivated in the future, for its better DX.

jsdanielh commented on web-client/src/lib.rs:
```diff
@@ -151,10 +162,46 @@ impl ClientConfiguration {
+    //     }
+    // }
+
+    pub fn build(&self) -> PlainClientConfigurationType {
```
Add `rustdoc` comments

jsdanielh commented on web-client/dist/bundler/worker.js:
```diff
@@ -0,0 +1,54 @@
+//         return Comlink.transferHandlers.get('proxy').serialize(obj);
+//     },
+//     // Cannot receive functions from worker
+// });
```
Maybe remove the commented code?

jsdanielh commented on web-client/dist/bundler/worker.js:
```diff
@@ -0,0 +1,54 @@
+    console.log('Initializing WASM worker');
+
+    const client = await Client.create(config);
+    // self.client = client; // Prevent garbage collection
```
Maybe remove the comment?

jsdanielh commented on web-client/dist/bundler/worker.js:
```diff
@@ -0,0 +1,54 @@
+    try {
+        await init(config);
+        self.postMessage({ ok: true });
+        // console.log('OK');
```
Maybe remove the comment?

jsdanielh commented on web-client/dist/bundler/worker.js:
```diff
@@ -0,0 +1,54 @@
+        // console.log('OK');
+    } catch (error) {
+        self.postMessage({ ok: false, error: error.message, stack: error.stack });
+        // console.error(error);
```
Maybe remove the comment?

jsdanielh commented on web-client/dist/transfer-handlers.js:
```diff
@@ -0,0 +1,17 @@
+    //     deserialize(port) {
+    //         return Comlink.transferHandlers.get('proxy').deserialize(port);
+    //     },
+    // });
```
Maybe remove the commented code?

jsdanielh commented on web-client/dist/web/worker.js:
```diff
@@ -0,0 +1,61 @@
+//         return Comlink.transferHandlers.get('proxy').serialize(obj);
+//     },
+//     // Cannot receive functions from worker
+// });
```
Maybe remove commented code?

jsdanielh commented on web-client/dist/web/worker.js:
```diff
@@ -0,0 +1,61 @@
+    await wasm_bindgen('./worker-wasm/index_bg.wasm');
+
+    const client = await Client.create(config);
+    // self.client = client; // Prevent garbage collection
```
Maybe remove commented code?

jsdanielh commented on web-client/dist/web/worker.js:
```diff
@@ -0,0 +1,61 @@
+    try {
+        await init(config);
+        self.postMessage({ ok: true });
+        // console.log('OK');
```
Maybe remove commented code?

jsdanielh commented on web-client/dist/web/worker.js:
```diff
@@ -0,0 +1,61 @@
+        // console.log('OK');
+    } catch (error) {
+        self.postMessage({ ok: false, error: error.message, stack: error.stack });
+        // console.error(error);
```
Maybe remove commented code?

jsdanielh commented on web-client/scripts/npm-publish.sh:
```diff
@@ -0,0 +1,5 @@
+./scripts/build-bundler.sh &&
+./scripts/build-web.sh &&
+cd dist &&
+npm publish --tag next &&
```
`npm publish` supports a folder argument. Can we use it instead of `cd dist` & `cd ..`

sisou replied:
I don't think it does. How would you pass a folder to `npm publish`? It never works for me.

jsdanielh commented on web-client/scripts/build-web.sh:
```diff
@@ -0,0 +1,8 @@
+wasm-pack build --weak-refs --target web --out-name index --out-dir dist/web/main-wasm &&
```
I'd suggest to use `set -e` at the beginning instead of `&&` for each line.

sisou replied:
Fixed by you in 67d08d40e67bfcfca7aa79de98b34a5891bc9271 already :+1:

jsdanielh commented on web-client/scripts/npm-publish.sh:
```diff
@@ -0,0 +1,5 @@
+./scripts/build-bundler.sh &&
```
I'd suggest to use `set -e` at the beginning instead of `&&` for each line.

jsdanielh commented on web-client/scripts/build-bundler.sh:
```diff
@@ -0,0 +1,8 @@
+wasm-pack build --weak-refs --target bundler --out-name index --out-dir dist/bundler/main-wasm &&
```
I'd suggest to use `set -e` at the beginning instead of `&&` for each line.

jsdanielh commented on web-client/dist/README.md:
```diff
@@ -3,7 +3,7 @@
 
 > **Note**
-> This web client is intended to be used in web browsers only, as it requires `window` to be available (no WASI support either).
+> This web client is intended to be used in web browsers only (no WASI support either). Other webworker-enabled environments are not yet targeted.
```
Maybe _supported_ instead of _targeted_?

sisou replied:
Ok.

jsdanielh commented on web-client/scripts/build-bundler.sh:
```diff
@@ -1,8 +1,8 @@
-wasm-pack build --weak-refs --target bundler --out-name index --out-dir dist/bundler/main-wasm &&
+wasm-pack build --weak-refs --target bundler --out-name index --out-dir dist/bundler/main-wasm -- --features "primitives" &&
```
Double quotes for `primitives` look to me unnecessary

jsdanielh commented on web-client/scripts/build-bundler.sh:
```diff
@@ -1,8 +1,8 @@
 rm dist/bundler/main-wasm/package.json &&
 rm dist/bundler/main-wasm/README.md &&
-wasm-pack build --weak-refs --target bundler --out-name index --out-dir dist/bundler/worker-wasm &&
+wasm-pack build --weak-refs --target bundler --out-name index --out-dir dist/bundler/worker-wasm -- --features "client" &&
```
Double quotes for `client` look to me unnecessary

jsdanielh commented on web-client/scripts/build-web.sh:
```diff
@@ -1,8 +1,8 @@
-wasm-pack build --weak-refs --target web --out-name index --out-dir dist/web/main-wasm &&
+wasm-pack build --weak-refs --target web --out-name index --out-dir dist/web/main-wasm -- --features "primitives" &&
```
Double quotes for `primitives` look to me unnecessary

jsdanielh commented on web-client/scripts/build-web.sh:
```diff
@@ -1,8 +1,8 @@
 rm dist/web/main-wasm/package.json &&
 rm dist/web/main-wasm/README.md &&
-wasm-pack build --weak-refs --target no-modules --out-name index --out-dir dist/web/worker-wasm &&
+wasm-pack build --weak-refs --target no-modules --out-name index --out-dir dist/web/worker-wasm -- --features "client" &&
```
Double quotes for `client` look to me unnecessary

jsdanielh commented on web-client/src/client_configuration.rs:
```diff
@@ -0,0 +1,129 @@
+/// This is a simplified version of the configuration that is used for regular nodes,
+/// since not all configuration knobs are available when running inside a browser.
+#[wasm_bindgen]
+pub struct ClientConfiguration {
```
Members are missing `rustdoc`

sisou replied:
The members' names are self-explanatory I would say.

jsdanielh commented on web-client/src/client_configuration.rs:
```diff
@@ -0,0 +1,129 @@
+
+#[derive(serde::Serialize, serde::Deserialize, Tsify)]
+#[serde(rename_all = "camelCase")]
+pub struct PlainClientConfiguration {
```
The `struct` and their members are missing `rustdoc`

sisou replied:
Again, self-explanatory.

jsdanielh commented on web-client/src/client_configuration.rs:
```diff
@@ -0,0 +1,129 @@
+    //         Ok(client) => client,
+    //         Err(_) => unreachable!(),
+    //     }
+    // }
```
Maybe remove commented code?

sisou replied:
Same as above:
> I am keeping it, but adding a TODO comment that explains that this method should be tried to be reactivated in the future, for its better DX.

jsdanielh commented on web-client/src/client_configuration.rs:
```diff
@@ -0,0 +1,129 @@
+    //     }
+    // }
+
+    pub fn build(&self) -> PlainClientConfigurationType {
```
Add `rustdoc` comments

sisou replied:
Duplicate comment.

jsdanielh commented on web-client/src/client.rs:
```diff
@@ -0,0 +1,1209 @@
+    config::config::ClientConfig,
+    config::config_file::{LogSettings, Seed},
+    extras::{panic::initialize_panic_reporting, web_logging::initialize_web_logging},
+};
```
All `nimiq` includes should be together

sisou replied:
They are. Do you mean I should remove the newline between `nimiq::` and `nimiq_*::` imports?

jsdanielh replied:
Yes, basically what you already did :+1:

jsdanielh commented on web-client/src/client.rs:
```diff
@@ -0,0 +1,1209 @@
+
+    /// Fetches the staker for the provided address from the network.
+    ///
+    /// Throws if the address cannot be parsed and on network errors.
```
Maybe _or_ on network errors.

sisou replied:
I think _and_ is correct.

jsdanielh commented on web-client/src/client.rs:
```diff
@@ -0,0 +1,1209 @@
+
+    /// Fetches the accounts for the provided addresses from the network.
+    ///
+    /// Throws if an address cannot be parsed and on network errors.
```
Maybe _or_ on network errors.

jsdanielh commented on web-client/src/client.rs:
```diff
@@ -0,0 +1,1209 @@
+
+    /// Fetches the stakers for the provided addresses from the network.
+    ///
+    /// Throws if an address cannot be parsed and on network errors.
```
Maybe _or_ on network errors.

jsdanielh commented on web-client/src/client.rs:
```diff
@@ -0,0 +1,1209 @@
+
+    /// Fetches the validator for the provided address from the network.
+    ///
+    /// Throws if the address cannot be parsed and on network errors.
```
Maybe _or_ on network errors.

jsdanielh commented on web-client/src/client.rs:
```diff
@@ -0,0 +1,1209 @@
+
+    /// Fetches the validators for the provided addresses from the network.
+    ///
+    /// Throws if an address cannot be parsed and on network errors.
```
Maybe _or_ on network errors.

jsdanielh commented on web-client/src/client.rs:
```diff
@@ -0,0 +1,1209 @@
+
+    /// Sends a transaction to the network and returns {@link PlainTransactionDetails}.
+    ///
+    /// Throws in case of a networking error.
```
Maybe _network_ error?

jsdanielh commented on web-client/src/client.rs:
```diff
@@ -0,0 +1,1209 @@
+            // If we got a transactions, return it
+            Ok(serde_wasm_bindgen::to_value(&details)?.into())
+        } else {
+            // If the transaction did not get included, return it as TransactionState::New
```
Maybe use backticks for `TransactionState::New`?

jsdanielh commented on web-client/src/client.rs:
```diff
@@ -0,0 +1,1209 @@
+
+    /// Fetches the transaction details for the given transaction hash.
+    #[wasm_bindgen(js_name = getTransaction)]
+    pub async fn get_transaction(
```
I'd move this function such that all `pub` functions are all together.

jsdanielh commented on web-client/src/client.rs:
```diff
@@ -0,0 +1,1209 @@
+    /// Up to a `limit` number of transaction receipts are returned from newest to oldest.
+    /// If the network does not have at least `min_peers` to query, then an error is returned.
+    #[wasm_bindgen(js_name = getTransactionReceiptsByAddress)]
+    pub async fn get_transaction_receipts_by_address(
```
I'd move this function such that all `pub` functions are all together.

jsdanielh commented on web-client/src/client.rs:
```diff
@@ -0,0 +1,1209 @@
+    /// Up to a `limit` number of transactions are returned from newest to oldest.
+    /// If the network does not have at least `min_peers` to query, then an error is returned.
+    #[wasm_bindgen(js_name = getTransactionsByAddress)]
+    pub async fn get_transactions_by_address(
```
I'd move this function such that all `pub` functions are all together.

sisou committed:
Implement Javascript files for example and package with updated client APIs

For the NPM package, we create two builds, one for bundlers (that can use import from package names), and one for plain web use, that needs to specify the exact file that is imported, both in the main thread and in the webworker.

This also updates the demo (now in example/) module.js file to use the updated client API, using the TransactionBuilder statically and passing serialized (to hex in this case) transactions when sending them. No other changes are required in the demo file.

The `Client` class/object defined in dist/client-proxy.js _shadows_ the `class Client` exported by the worker types and adheres to that one's static API. For the user it is transparent.

sisou committed:
Build types separately to merge all into one definition file

jsdanielh committed:
web-client: Fix client log level setting

Fix the setting of the log level for the client. The `primitives`
feature was missing the `camelCase` `serde` directive and the `Tsify`
derive directive for the `ClientConfiguration` struct.
Also made the `primitives` and `client` features the default features
and changed the script to disable the default features.
Also changes the script to use `set -e` to avoid the usage of `&&` in
each line.

jsdanielh committed:
web-client: Add handler for the `visible` event

Add handler to trigger the check of peers on the `visible` event from
the `visibilitychange` event list from the document.

sisou committed:
Allow passing class instances to the client

By transparently serializing them to their plain format before being sent to the worker.

sisou committed:
Cleanup, remove commented code, wording, docs, move pub methods together

jsdanielh commented:
Rebasing it to get it merged
